task/task.go

```diff
@@ -47,7 +59,11 @@ func DoRetryWithTimeout(t func() (interface{}, bool, error), timeout, timeBefore
 	case <-done:
 		return out, err
 	case <-time.After(timeout):
+		errLock.Lock()
+		defer errLock.Unlock()
 		quit <- true
-		return out, ErrTimedOut
+		return out, &ErrTimedOut{
+			Reason: err.Error(),
```

Comment:
Check if err is nil before calling err.Error()
There is still a possibility where the above goroutine succeeded and then released the lock and this code got it to return ErrTimedOut. 

To be safe, in addition to the above check for nil, we should do an errLock.Unlock() only after `true` is sent over the done channel.

Reply:
fixed